src/pulsar_producer.erl

```diff
@@ -355,19 +348,39 @@ do_connect(State = #{opts := Opts, broker_server := {Host, Port}}) ->
              [{state_timeout, ?RECONNECT_TIMEOUT, do_connect}]}
     end.
 
-code_change({down, _Vsn} = Direction, State, Data0, _Extra) ->
+code_change({down, _ToVsn}, State, Data, Extra) when is_map(Data) ->
+    #{to_version := ToVsn} = Extra,
+    case pulsar_relup:is_before_replayq(ToVsn) of
+        true ->
+            do_code_change_down_replayq(State, Data, Extra);
+        false ->
+            {ok, State, Data}
+    end;
+code_change({down, _Vsn}, State, Data, _Extra) ->
+    {ok, State, Data};
+code_change(_ToVsn, State, DataRec, Extra) when is_tuple(DataRec) ->
+    #{to_version := ToVsn} = Extra,
+    case pulsar_relup:is_before_replayq(ToVsn) of
+        true ->
+            {ok, State, DataRec};
```

Comment:
this should never happen?

Reply:
No.  I was just being conservative here.